yjit/src/codegen.rs

```diff
@@ -4460,28 +4459,42 @@ fn jit_rb_int_lshift(
     }
     guard_two_fixnums(jit, asm, ocb);
 
+    let comptime_shift = jit.peek_at_stack(&asm.ctx, 0);
+
+    if !comptime_shift.fixnum_p() {
+        return false;
+    }
+
+    // Untag the fixnum shift amount
+    let shift_amt = comptime_shift.as_isize() >> 1;
+
+    if shift_amt > 63 || shift_amt < 0 {
+        return false;
+    }
+
     let rhs = asm.stack_pop(1);
     let lhs = asm.stack_pop(1);
 
-    // Ruby supports using a negative shift value
-    asm.comment("Guard shift negative");
-    let shift_val = asm.sub(rhs, 1.into());
-    asm.cmp(shift_val, 0.into());
-    asm.jl(Target::side_exit(Counter::lshift_range));
+    // Guard on the shift value we speculated on
+    asm.cmp(rhs, comptime_shift.into());
+    asm.jne(Target::side_exit(Counter::lshift_amt_changed));
```

Comment:
I wonder if it's useful to have a chain guard here, but I guess we'd rather like to save memory, assuming the shift amount is always an integer literal?

Reply:
It seems in optcarrot it is always constant. It's kind of rare to see people using bitwise shifts with a variable amount.